File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/MemoizeFbtAndMacroOperandsInSameScope.ts
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  HIRFunction,
  IdentifierId,
  InstructionValue,
  makeInstructionId,
  MutableRange,
  Place,
  ReactiveScope,
} from '../HIR';
import {Macro} from '../HIR/Environment';
import {eachInstructionValueOperand} from '../HIR/visitors';

/**
 * Whether a macro requires its arguments to be transitively inlined (eg fbt)
 * or just avoid having the top-level values be converted to variables (eg fbt.param)
 */
enum InlineLevel {
  Transitive = 'Transitive',
  Shallow = 'Shallow',
}
type MacroDefinition = {
  level: InlineLevel;
  properties: Map<string, MacroDefinition> | null;
};

const SHALLOW_MACRO: MacroDefinition = {
  level: InlineLevel.Shallow,
  properties: null,
};
const TRANSITIVE_MACRO: MacroDefinition = {
  level: InlineLevel.Transitive,
  properties: null,
};
const FBT_MACRO: MacroDefinition = {
  level: InlineLevel.Transitive,
  properties: new Map([['*', SHALLOW_MACRO]]),
};
FBT_MACRO.properties!.set('enum', FBT_MACRO);

/**
 * This pass supports the `fbt` translation system (https://facebook.github.io/fbt/)
 * as well as similar user-configurable macro-like APIs where it's important that
 * the name of the function not be changed, and it's literal arguments not be turned
 * into temporaries.
 *
 * ## FBT
 *
 * FBT provides the `<fbt>` JSX element and `fbt()` calls (which take params in the
 * form of `<fbt:param>` children or `fbt.param()` arguments, respectively). These
 * tags/functions have restrictions on what types of syntax may appear as props/children/
 * arguments, notably that variable references may not appear directly — variables
 * must always be wrapped in a `<fbt:param>` or `fbt.param()`.
 *
 * To ensure that Forget doesn't rewrite code to violate this restriction, we force
 * operands to fbt tags/calls have the same scope as the tag/call itself.
 *
 * Note that this still allows the props/arguments of `<fbt:param>`/`fbt.param()`
 * to be independently memoized.
 *
 * ## User-defined macro-like function
 *
 * Users can also specify their own functions to be treated similarly to fbt via the
 * `customMacros` environment configuration. By default, user-supplied custom macros
 * have their arguments transitively inlined.
 */
export function memoizeFbtAndMacroOperandsInSameScope(
  fn: HIRFunction,
): Set<IdentifierId> {
  const macroKinds = new Map<Macro, MacroDefinition>([
    ...Array.from(FBT_TAGS.entries()),
    ...(fn.env.config.customMacros ?? []).map(
      name => [name, TRANSITIVE_MACRO] as [Macro, MacroDefinition],
    ),
  ]);
  /**
   * Forward data-flow analysis to identify all macro tags, including
   * things like `fbt.foo.bar(...)`
   */
  const macroTags = populateMacroTags(fn, macroKinds);

  /**
   * Reverse data-flow analysis to merge arguments to macro *invocations*
   * based on the kind of the macro
   */
  const macroValues = mergeMacroArguments(fn, macroTags, macroKinds);

  return macroValues;
}

const FBT_TAGS: Map<string, MacroDefinition> = new Map([
  ['fbt', FBT_MACRO],
  ['fbt:param', SHALLOW_MACRO],
  ['fbt:enum', FBT_MACRO],
  ['fbt:plural', SHALLOW_MACRO],
  ['fbs', FBT_MACRO],
  ['fbs:param', SHALLOW_MACRO],
  ['fbs:enum', FBT_MACRO],
  ['fbs:plural', SHALLOW_MACRO],
]);
export const SINGLE_CHILD_FBT_TAGS: Set<string> = new Set([
  'fbt:param',
  'fbs:param',
]);

function populateMacroTags(
  fn: HIRFunction,
  macroKinds: Map<Macro, MacroDefinition>,
): Map<IdentifierId, MacroDefinition> {
  const macroTags = new Map<IdentifierId, MacroDefinition>();
  for (const block of fn.body.blocks.values()) {
    for (const instr of block.instructions) {
      const {lvalue, value} = instr;
      switch (value.kind) {
        case 'Primitive': {
          if (typeof value.value === 'string') {
            const macroDefinition = macroKinds.get(value.value);
            if (macroDefinition != null) {
              /*
               * We don't distinguish between tag names and strings, so record
               * all `fbt` string literals in case they are used as a jsx tag.
               */
              macroTags.set(lvalue.identifier.id, macroDefinition);
            }
          }
          break;
        }
        case 'LoadGlobal': {
          let macroDefinition = macroKinds.get(value.binding.name);
          if (macroDefinition != null) {
            macroTags.set(lvalue.identifier.id, macroDefinition);
          }
          break;
        }
        case 'PropertyLoad': {
          if (typeof value.property === 'string') {
            const macroDefinition = macroTags.get(value.object.identifier.id);
            if (macroDefinition != null) {
              const propertyDefinition =
                macroDefinition.properties != null
                  ? (macroDefinition.properties.get(value.property) ??
                    macroDefinition.properties.get('*'))
                  : null;
              const propertyMacro = propertyDefinition ?? macroDefinition;
              macroTags.set(lvalue.identifier.id, propertyMacro);
            }
          }
          break;
        }
      }
    }
  }
  return macroTags;
}

function mergeMacroArguments(
  fn: HIRFunction,
  macroTags: Map<IdentifierId, MacroDefinition>,
  macroKinds: Map<Macro, MacroDefinition>,
): Set<IdentifierId> {
  const macroValues = new Set<IdentifierId>(macroTags.keys());
  for (const block of Array.from(fn.body.blocks.values()).reverse()) {
    for (let i = block.instructions.length - 1; i >= 0; i--) {
      const instr = block.instructions[i]!;
      const {lvalue, value} = instr;
      switch (value.kind) {
        case 'DeclareContext':
        case 'DeclareLocal':
        case 'Destructure':
        case 'LoadContext':
        case 'LoadLocal':
        case 'PostfixUpdate':
        case 'PrefixUpdate':
        case 'StoreContext':
        case 'StoreLocal': {
          // Instructions that never need to be merged
          break;
        }
        case 'CallExpression':
        case 'MethodCall': {
          const scope = lvalue.identifier.scope;
          if (scope == null) {
            continue;
          }
          const callee =
            value.kind === 'CallExpression' ? value.callee : value.property;
          const macroDefinition =
            macroTags.get(callee.identifier.id) ??
            macroTags.get(lvalue.identifier.id);
          if (macroDefinition != null) {
            visitOperands(
              macroDefinition,
              scope,
              lvalue,
              value,
              macroValues,
              macroTags,
            );
          }
          break;
        }
        case 'JsxExpression': {
          const scope = lvalue.identifier.scope;
          if (scope == null) {
            continue;
          }
          let macroDefinition;
          if (value.tag.kind === 'Identifier') {
            macroDefinition = macroTags.get(value.tag.identifier.id);
          } else {
            macroDefinition = macroKinds.get(value.tag.name);
          }
          macroDefinition ??= macroTags.get(lvalue.identifier.id);
          if (macroDefinition != null) {
            visitOperands(
              macroDefinition,
              scope,
              lvalue,
              value,
              macroValues,
              macroTags,
            );
          }
          break;
        }
        default: {
          const scope = lvalue.identifier.scope;
          if (scope == null) {
            continue;
          }
          const macroDefinition = macroTags.get(lvalue.identifier.id);
          if (macroDefinition != null) {
            visitOperands(
              macroDefinition,
              scope,
              lvalue,
              value,
              macroValues,
              macroTags,
            );
          }
          break;
        }
      }
    }
    for (const phi of block.phis) {
      const scope = phi.place.identifier.scope;
      if (scope == null) {
        continue;
      }
      const macroDefinition = macroTags.get(phi.place.identifier.id);
      if (
        macroDefinition == null ||
        macroDefinition.level === InlineLevel.Shallow
      ) {
        continue;
      }
      macroValues.add(phi.place.identifier.id);
      for (const operand of phi.operands.values()) {
        operand.identifier.scope = scope;
        expandFbtScopeRange(scope.range, operand.identifier.mutableRange);
        macroTags.set(operand.identifier.id, macroDefinition);
        macroValues.add(operand.identifier.id);
      }
    }
  }
  return macroValues;
}

function expandFbtScopeRange(
  fbtRange: MutableRange,
  extendWith: MutableRange,
): void {
  if (extendWith.start !== 0) {
    fbtRange.start = makeInstructionId(
      Math.min(fbtRange.start, extendWith.start),
    );
  }
}

function visitOperands(
  macroDefinition: MacroDefinition,
  scope: ReactiveScope,
  lvalue: Place,
  value: InstructionValue,
  macroValues: Set<IdentifierId>,
  macroTags: Map<IdentifierId, MacroDefinition>,
): void {
  macroValues.add(lvalue.identifier.id);
  for (const operand of eachInstructionValueOperand(value)) {
    if (macroDefinition.level === InlineLevel.Transitive) {
      operand.identifier.scope = scope;
      expandFbtScopeRange(scope.range, operand.identifier.mutableRange);
      macroTags.set(operand.identifier.id, macroDefinition);
    }
    macroValues.add(operand.identifier.id);
  }
}

File: compiler/crates/react_compiler_inference/src/memoize_fbt_and_macro_operands_in_same_scope.rs
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

//! Port of MemoizeFbtAndMacroOperandsInSameScope from TypeScript.
//!
//! Ensures that FBT (Facebook Translation) expressions and their operands
//! are memoized within the same reactive scope. Also supports user-configured
//! custom macro-like APIs via `customMacros` configuration.
//!
//! The pass has two phases:
//! 1. Forward data-flow: identify all macro tags (including property loads like `fbt.param`)
//! 2. Reverse data-flow: merge arguments of macro invocations into the same scope

use std::collections::{HashMap, HashSet};

use react_compiler_hir::environment::Environment;
use react_compiler_hir::visitors;
use react_compiler_hir::{
    HirFunction, IdentifierId, InstructionValue, JsxTag,
    PrimitiveValue, PropertyLiteral, ScopeId,
};

/// Whether a macro requires its arguments to be transitively inlined (e.g., fbt)
/// or just avoids having the top-level values be converted to variables (e.g., fbt.param).
#[derive(Debug, Clone)]
enum InlineLevel {
    Transitive,
    Shallow,
}

/// Defines how a macro and its properties should be handled.
#[derive(Debug, Clone)]
struct MacroDefinition {
    level: InlineLevel,
    /// Maps property names to their own MacroDefinition. `"*"` is a wildcard.
    properties: Option<HashMap<String, MacroDefinition>>,
}

fn shallow_macro() -> MacroDefinition {
    MacroDefinition {
        level: InlineLevel::Shallow,
        properties: None,
    }
}

fn transitive_macro() -> MacroDefinition {
    MacroDefinition {
        level: InlineLevel::Transitive,
        properties: None,
    }
}

fn fbt_macro() -> MacroDefinition {
    let mut props = HashMap::new();
    props.insert("*".to_string(), shallow_macro());
    // fbt.enum gets FBT_MACRO (recursive/transitive)
    // We'll fill this in after construction since it's self-referential.
    // Instead, we use a special marker and handle it in property lookup.
    let mut fbt = MacroDefinition {
        level: InlineLevel::Transitive,
        properties: Some(props),
    };
    // Add "enum" as a recursive reference (same as FBT_MACRO)
    // Since we can't do self-referential structs, we clone the structure.
    let enum_macro = MacroDefinition {
        level: InlineLevel::Transitive,
        properties: Some({
            let mut p = HashMap::new();
            p.insert("*".to_string(), shallow_macro());
            // enum's enum is also recursive, but in practice the depth is bounded
            p.insert("enum".to_string(), transitive_macro());
            p
        }),
    };
    fbt.properties.as_mut().unwrap().insert("enum".to_string(), enum_macro);
    fbt
}

/// Built-in FBT tags and their macro definitions.
fn fbt_tags() -> HashMap<String, MacroDefinition> {
    let mut tags = HashMap::new();
    tags.insert("fbt".to_string(), fbt_macro());
    tags.insert("fbt:param".to_string(), shallow_macro());
    tags.insert("fbt:enum".to_string(), fbt_macro());
    tags.insert("fbt:plural".to_string(), shallow_macro());
    tags.insert("fbs".to_string(), fbt_macro());
    tags.insert("fbs:param".to_string(), shallow_macro());
    tags.insert("fbs:enum".to_string(), fbt_macro());
    tags.insert("fbs:plural".to_string(), shallow_macro());
    tags
}

/// Main entry point. Returns the set of identifier IDs that are fbt/macro operands.
pub fn memoize_fbt_and_macro_operands_in_same_scope(
    func: &HirFunction,
    env: &mut Environment,
) -> HashSet<IdentifierId> {
    // Phase 1: Build macro kinds map from built-in FBT tags + custom macros
    let mut macro_kinds: HashMap<String, MacroDefinition> = fbt_tags();
    if let Some(ref custom_macros) = env.config.custom_macros {
        for name in custom_macros {
            macro_kinds.insert(name.clone(), transitive_macro());
        }
    }

    // Phase 2: Forward data-flow to identify all macro tags
    let mut macro_tags = populate_macro_tags(func, &macro_kinds);

    // Phase 3: Reverse data-flow to merge arguments of macro invocations
    let macro_values = merge_macro_arguments(func, env, &mut macro_tags, &macro_kinds);

    macro_values
}

/// Forward data-flow analysis to identify all macro tags, including
/// things like `fbt.foo.bar(...)`.
fn populate_macro_tags(
    func: &HirFunction,
    macro_kinds: &HashMap<String, MacroDefinition>,
) -> HashMap<IdentifierId, MacroDefinition> {
    let mut macro_tags: HashMap<IdentifierId, MacroDefinition> = HashMap::new();

    for block in func.body.blocks.values() {
        for &instr_id in &block.instructions {
            let instr = &func.instructions[instr_id.0 as usize];
            let lvalue_id = instr.lvalue.identifier;

            match &instr.value {
                InstructionValue::Primitive {
                    value: PrimitiveValue::String(s),
                    ..
                } => {
                    if let Some(macro_def) = macro_kinds.get(s.as_str()) {
                        // We don't distinguish between tag names and strings, so record
                        // all `fbt` string literals in case they are used as a jsx tag.
                        macro_tags.insert(lvalue_id, macro_def.clone());
                    }
                }
                InstructionValue::LoadGlobal { binding, .. } => {
                    let name = binding.name();
                    if let Some(macro_def) = macro_kinds.get(name) {
                        macro_tags.insert(lvalue_id, macro_def.clone());
                    }
                }
                InstructionValue::PropertyLoad {
                    object, property, ..
                } => {
                    if let PropertyLiteral::String(prop_name) = property {
                        if let Some(macro_def) = macro_tags.get(&object.identifier).cloned() {
                            let property_macro = if let Some(ref props) = macro_def.properties {
                                let prop_def = props
                                    .get(prop_name.as_str())
                                    .or_else(|| props.get("*"));
                                match prop_def {
                                    Some(def) => def.clone(),
                                    None => macro_def.clone(),
                                }
                            } else {
                                macro_def.clone()
                            };
                            macro_tags.insert(lvalue_id, property_macro);
                        }
                    }
                }
                _ => {}
            }
        }
    }

    macro_tags
}

/// Reverse data-flow analysis to merge arguments to macro *invocations*
/// based on the kind of the macro.
fn merge_macro_arguments(
    func: &HirFunction,
    env: &mut Environment,
    macro_tags: &mut HashMap<IdentifierId, MacroDefinition>,
    macro_kinds: &HashMap<String, MacroDefinition>,
) -> HashSet<IdentifierId> {
    let mut macro_values: HashSet<IdentifierId> = macro_tags.keys().copied().collect();

    // Iterate blocks in reverse order
    let block_ids: Vec<_> = func.body.blocks.keys().copied().collect();
    for &block_id in block_ids.iter().rev() {
        let block = &func.body.blocks[&block_id];

        // Iterate instructions in reverse order
        for &instr_id in block.instructions.iter().rev() {
            let instr = &func.instructions[instr_id.0 as usize];
            let lvalue_id = instr.lvalue.identifier;

            match &instr.value {
                // Instructions that never need to be merged
                InstructionValue::DeclareContext { .. }
                | InstructionValue::DeclareLocal { .. }
                | InstructionValue::Destructure { .. }
                | InstructionValue::LoadContext { .. }
                | InstructionValue::LoadLocal { .. }
                | InstructionValue::PostfixUpdate { .. }
                | InstructionValue::PrefixUpdate { .. }
                | InstructionValue::StoreContext { .. }
                | InstructionValue::StoreLocal { .. } => {
                    // Skip these
                }

                InstructionValue::CallExpression { callee, .. }
                | InstructionValue::MethodCall {
                    property: callee, ..
                } => {
                    let scope_id = match env.identifiers[lvalue_id.0 as usize].scope {
                        Some(s) => s,
                        None => continue,
                    };

                    let macro_def = macro_tags
                        .get(&callee.identifier)
                        .or_else(|| macro_tags.get(&lvalue_id))
                        .cloned();

                    if let Some(macro_def) = macro_def {
                        visit_operands(
                            &macro_def,
                            scope_id,
                            lvalue_id,
                            &instr.value,
                            env,
                            &mut macro_values,
                            macro_tags,
                        );
                    }
                }

                InstructionValue::JsxExpression { tag, .. } => {
                    let scope_id = match env.identifiers[lvalue_id.0 as usize].scope {
                        Some(s) => s,
                        None => continue,
                    };

                    let macro_def = match tag {
                        JsxTag::Place(place) => {
                            macro_tags.get(&place.identifier).cloned()
                        }
                        JsxTag::Builtin(builtin) => {
                            macro_kinds.get(builtin.name.as_str()).cloned()
                        }
                    };

                    let macro_def = macro_def
                        .or_else(|| macro_tags.get(&lvalue_id).cloned());

                    if let Some(macro_def) = macro_def {
                        visit_operands(
                            &macro_def,
                            scope_id,
                            lvalue_id,
                            &instr.value,
                            env,
                            &mut macro_values,
                            macro_tags,
                        );
                    }
                }

                // Default case: check if lvalue is a macro tag
                _ => {
                    let scope_id = match env.identifiers[lvalue_id.0 as usize].scope {
                        Some(s) => s,
                        None => continue,
                    };

                    let macro_def = macro_tags.get(&lvalue_id).cloned();
                    if let Some(macro_def) = macro_def {
                        visit_operands(
                            &macro_def,
                            scope_id,
                            lvalue_id,
                            &instr.value,
                            env,
                            &mut macro_values,
                            macro_tags,
                        );
                    }
                }
            }
        }

        // Handle phis
        let block = &func.body.blocks[&block_id];
        for phi in &block.phis {
            let scope_id = match env.identifiers[phi.place.identifier.0 as usize].scope {
                Some(s) => s,
                None => continue,
            };

            let macro_def = match macro_tags.get(&phi.place.identifier).cloned() {
                Some(def) => def,
                None => continue,
            };

            if matches!(macro_def.level, InlineLevel::Shallow) {
                continue;
            }

            macro_values.insert(phi.place.identifier);

            // Collect operand updates to avoid borrow issues
            let operand_updates: Vec<(IdentifierId, MacroDefinition)> = phi
                .operands
                .values()
                .map(|operand| (operand.identifier, macro_def.clone()))
                .collect();

            for (operand_id, def) in operand_updates {
                env.identifiers[operand_id.0 as usize].scope = Some(scope_id);
                expand_fbt_scope_range(env, scope_id, operand_id);
                macro_tags.insert(operand_id, def);
                macro_values.insert(operand_id);
            }
        }
    }

    macro_values
}

/// Expand the scope range on the environment, reading from identifier's mutable_range.
/// Also syncs mutable_range for all identifiers in this scope whose range matched
/// the old scope range. In TS, identifier.mutableRange shares the same object as
/// scope.range, so mutations are automatically visible; in Rust we must propagate.
/// Equivalent to TS `expandFbtScopeRange`.
fn expand_fbt_scope_range(env: &mut Environment, scope_id: ScopeId, operand_id: IdentifierId) {
    let extend_start = env.identifiers[operand_id.0 as usize].mutable_range.start;
    if extend_start.0 == 0 {
        return;
    }
    let old_range = env.scopes[scope_id.0 as usize].range.clone();
    let new_start = old_range.start.0.min(extend_start.0);
    if new_start == old_range.start.0 {
        return;
    }
    env.scopes[scope_id.0 as usize].range.start.0 = new_start;
    let new_range = env.scopes[scope_id.0 as usize].range.clone();
    for ident in &mut env.identifiers {
        if ident.scope == Some(scope_id)
            && ident.mutable_range.start == old_range.start
            && ident.mutable_range.end == old_range.end
        {
            ident.mutable_range = new_range.clone();
        }
    }
}

/// Visit operands for an instruction value, merging them into the same scope
/// if the macro definition requires transitive inlining.
fn visit_operands(
    macro_def: &MacroDefinition,
    scope_id: ScopeId,
    lvalue_id: IdentifierId,
    value: &InstructionValue,
    env: &mut Environment,
    macro_values: &mut HashSet<IdentifierId>,
    macro_tags: &mut HashMap<IdentifierId, MacroDefinition>,
) {
    macro_values.insert(lvalue_id);

    let operand_ids: Vec<IdentifierId> =
        visitors::each_instruction_value_operand_with_functions(value, &env.functions)
            .into_iter()
            .map(|p| p.identifier)
            .collect();

    for operand_id in operand_ids {
        if matches!(macro_def.level, InlineLevel::Transitive) {
            env.identifiers[operand_id.0 as usize].scope = Some(scope_id);
            expand_fbt_scope_range(env, scope_id, operand_id);
            macro_tags.insert(operand_id, macro_def.clone());
        }
        macro_values.insert(operand_id);
    }
}

